// mediatr-ts/lib/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Handler = exports.pipelineBehavior = exports.Mediator = exports.notificationHandler = exports.NotificationData = exports.requestHandler = exports.RequestData = void 0;
var requestData_js_1 = __importDefault(require("./models/requestData.js"));
exports.RequestData = requestData_js_1.default;
var requestHandler_attribute_js_1 = __importDefault(require("./attributes/requestHandler.attribute.js"));
exports.requestHandler = requestHandler_attribute_js_1.default;
var notificationData_js_1 = __importDefault(require("./models/notificationData.js"));
exports.NotificationData = notificationData_js_1.default;
var notificationHandler_attribute_js_1 = __importDefault(require("./attributes/notificationHandler.attribute.js"));
exports.notificationHandler = notificationHandler_attribute_js_1.default;
var pipelineBehavior_attribute_js_1 = __importDefault(require("./attributes/pipelineBehavior.attribute.js"));
exports.pipelineBehavior = pipelineBehavior_attribute_js_1.default;
var mediator_1 = __importDefault(require("./models/mediator"));
exports.Mediator = mediator_1.default;
/**
 * @deprecated Use requestHandler instead
 */
exports.Handler = requestHandler_attribute_js_1.default;
